File: src/main/java/com/2u/synthetic-testing/DatadogClient.java
import java.net.HttpURLConnection;
import java.net.URL;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.concurrent.TimeUnit;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class DatadogClient {
    private static final String DATADOG_SYNTHETIC_TESTS_API_URL = "https://api.datadoghq.com/api/v1/synthetics/tests";

    private static final long MAX_ALLOWABLE_TIME_MS = 60 * 1000; // 1 minute

    private String apiKey;
    private String appKey;

    public DatadogClient(String apiKey, String appKey) {
        this.apiKey = apiKey;
        this.appKey = appKey;
    }

    /* ******************* Datadog API actions *****************/

    public String triggerSyntheticTests(SyntheticTestRequest[] testRequests) throws Exception {
        /*
        ** Trigger a synthetic test run consisting of one or more tests. At best this queues a request; the actual run
        ** will occur later, if at all. The request will just be dropped if testing is paused
        ** at the time of the request
        **
        ** Returns the testRunId for the requested test run
        */
        String url = DATADOG_SYNTHETIC_TESTS_API_URL + "/trigger";
        URL obj = new URL(url);
        HttpURLConnection con = (HttpURLConnection) obj.openConnection();

        // Set the request method and headers
        System.out.println("API Key: " + apiKey);
        System.out.println("Application Key: " + appKey);
        con.setRequestMethod("POST");
        con.setRequestProperty("Content-Type", "application/json");
        con.setRequestProperty("DD-API-KEY", apiKey);
        con.setRequestProperty("DD-APPLICATION-KEY", appKey);

        // Send the request with a request body indicating the specific tests to be run
        String jsonRequestBody = String.format("{\"tests\":[{\"public_id\":\"%s\"}]}", testRequests[0].testId);
        con.setDoOutput(true);
        try (OutputStream os = con.getOutputStream()) {
            byte[] requestBody = jsonRequestBody.getBytes("utf-8");
            os.write(requestBody, 0, requestBody.length);
        }

        // Get the response
        int responseCode = con.getResponseCode();
        System.out.println("POST Response Code :: " + responseCode);
        if (responseCode != 200) {
            throw new Exception("Datadog API error");
        }
        String responseBody = DatadogClient.getResponseBody(con).toString();

        // We only care about the test run ID in the response, which we'll use later on
        // to interrogate for test results. Extract and return it
        return DatadogClient.jsonGetTestRunId(responseBody.toString());
    }

    public String getTestingResults(String testRunId, SyntheticTestRequest[] testRequests) throws Exception {
        String jsonTestRunResults = null;
        for (SyntheticTestRequest request : testRequests) {
            jsonTestRunResults = waitForTestOnTestRunResult(testRunId, request.testId);
        }

        return jsonTestRunResults; // TODO: Create and return array of results; don't just return the last one
    }

   // The datadog API doesn't return the results for all tests on a test run; rather, it returns the results on a test
   // across all test runs that have run the test. This forces us to make extra calls
   //
   // TBD: see whether there's another Datadog API call that actually does what we want
   private String waitForTestOnTestRunResult(String testRunId, String testId) throws Exception {
        /*
        ** Poll for test run completion subject to a maximum allowable time
        **
        ** Returns the json structure with all test run results for tests in this test run
        */
        String jsonTestOnTestRunResult = null;
        long startTime = System.currentTimeMillis();
        while (jsonTestOnTestRunResult == null && (System.currentTimeMillis() - startTime) < MAX_ALLOWABLE_TIME_MS) {
            TimeUnit.SECONDS.sleep(5); // Poll every 5 seconds
            jsonTestOnTestRunResult = getJsonTestOnTestRunResult(testRunId, testId);
        }

        if (jsonTestOnTestRunResult == null) {
            throw new Exception("The test run timed out.");
        }
        return jsonTestOnTestRunResult;
    }

    private String getJsonTestOnTestRunResult(String testRunId, String testId) throws Exception {
        /*
        ** This method is called in a polling loop and returns a null response if the test run
        ** has not yet completed; otherwise, it returns a json structure with all test results in it
        */

        String url = DATADOG_SYNTHETIC_TESTS_API_URL + "/" + testId + "/results";
        URL obj = new URL(url);
        HttpURLConnection con = (HttpURLConnection) obj.openConnection();

        // Set the request method and headers. We don't specify which test we care about, as this
        // Datadog API returns the results for all tests run in the test run
        con.setRequestMethod("GET");
        con.setRequestProperty("DD-API-KEY", apiKey);
        con.setRequestProperty("DD-APPLICATION-KEY", appKey);

        // Read the response
        int responseCode = con.getResponseCode();
        System.out.println("GET Response Code :: " + responseCode);
        if (responseCode != 200) {
            return null;
        }
        String responseBody = DatadogClient.getResponseBody(con);
        String testResult = DatadogClient.jsonGetSecondFieldValueAfterFirstField(
            responseBody,"result_id", "passed\":", testRunId);
        return testResult;
    }

    /* *************** HTTP connection utilities *****************/

    private static String getResponseBody(HttpURLConnection con) throws Exception {
        BufferedReader in = new BufferedReader(new InputStreamReader(con.getInputStream()));
        String responseLine;
        StringBuilder responseBody = new StringBuilder();
        while ((responseLine = in.readLine()) != null) {
            responseBody.append(responseLine);
        }
        in.close();

        return responseBody.toString();
    }

    /* *************** JSON parsing **************** */

    private static String jsonGetTestRunId(String jsonResponse) throws Exception {
        // Extract test run ID from trigger test run result. Appears as result_id in response.
        // Example JSON response format:
        // { "result_id": "12345" }

        String key = "\"result_id\":\"";
        int index = jsonResponse.indexOf(key);
        if (index != -1) {
            int startIndex = index + key.length();
            int endIndex = jsonResponse.indexOf("\"", startIndex);
            return jsonResponse.substring(startIndex, endIndex);
        }
        throw new Exception("Trigger request failed");
    }

    private static String jsonGetSecondFieldValueAfterFirstField(
        String input,
        String testRunIdTag,
        String testResultTag,
        String testRunId) throws Exception {

        String regex = testRunIdTag + "\\s*([^,]+).*?" + testResultTag + "\\s*([^,]+)";

        Pattern pattern = Pattern.compile(regex);
        Matcher matcher = pattern.matcher(input);

        String testRunIdValue = null;
        String testResultValue = null;
        while (testRunIdValue == null && matcher.find()) {
            testRunIdValue = matcher.group(1).trim();
            testResultValue = matcher.group(2).trim();

            if (testRunIdValue == testRunId) {
                break;
            }
        }

        if (testResultValue == null) {
            throw new Exception("Failed to find test result in test run");
        }

        return testResultValue;
    }
}

File: datadog_client.py
import json
import requests
import time
import re


class DatadogClient:
    DATADOG_SYNTHETIC_TESTS_API_URL = "https://api.datadoghq.com/api/v1/synthetics/tests"
    MAX_ALLOWABLE_TIME_SECS = 60 # 1 minute

    def __init__(self, api_key, app_key):
        self.api_key = api_key
        self.app_key = app_key

    def trigger_synthetic_tests(self, test_requests):
        """
        Trigger a synthetic test run for one or more tests.
        Returns the testRunId for the requested test run.
        """
        url = f"{self.DATADOG_SYNTHETIC_TESTS_API_URL}/trigger"
        headers = {
            "Content-Type": "application/json",
            "DD-API-KEY": self.api_key,
            "DD-APPLICATION-KEY": self.app_key
        }
        json_request_body = {"tests": [{"public_id": test_requests[0].test_id}]}

        response = requests.post(url, headers=headers, json=json_request_body)

        if response.status_code != 200:
            raise Exception("Datadog API error")

        response_body = response.json()
        result = response_body['results'][0]
        test_run_id = result['result_id']
        return test_run_id

    def get_testing_results(self, test_run_id, test_requests):
        json_test_run_results = None
        for request in test_requests:
            json_test_run_results = self.wait_for_test_on_test_run_result(test_run_id, request.test_id)

        return json_test_run_results  # TODO: Create and return array of results; don't just return the last one

    def wait_for_test_on_test_run_result(self, test_run_id, test_id):
        """
        Poll for test run completion within the maximum allowable time.
        Returns the JSON structure with all test run results.
        """
        start_time = time.time()
        json_test_on_test_run_result = None
        while json_test_on_test_run_result is None and (time.time() - start_time) < (self.MAX_ALLOWABLE_TIME_SECS):
            time.sleep(5)  # Poll every 5 seconds
            print("Testing on DD result")
            json_test_on_test_run_result = self.get_json_test_on_test_run_result(test_run_id, test_id)

        if json_test_on_test_run_result is None:
            raise Exception("The test run timed out.")

        return json_test_on_test_run_result

    def get_json_test_on_test_run_result(self, test_run_id, test_id):
        """
        Called in a polling loop; returns JSON structure if test run has completed.
        """
        url = f"{self.DATADOG_SYNTHETIC_TESTS_API_URL}/{test_id}/results"
        headers = {
            "DD-API-KEY": self.api_key,
            "DD-APPLICATION-KEY": self.app_key
        }

        response = requests.get(url, headers=headers)

        if response.status_code != 200:
            return None

        response_json = response.json()
        return self.get_pass_fail_result(response_json, test_run_id)

    @staticmethod
    def json_get_test_run_id(json_response):
        """
        Extract test run ID from trigger test run result.
        """
        key = "result_id"
        if key in json_response:
            return json_response[key]
        raise Exception("Trigger request failed")

    @staticmethod
    def get_pass_fail_result(input_json, test_run_id):
        """
        Extracts pass/fail result from the specific test run in the aggregate test run result set
        """
        aggregate_results = input_json['results']
        result = [r for r in aggregate_results if r['result_id'] == test_run_id]
        if result is not None:
            test_run_data = result[0]['result']
            pass_fail = test_run_data['passed']
            return pass_fail

        raise Exception("Failed to find test result in test run")
